graphql/src/execution/execution.rs

```diff
@@ -67,6 +74,151 @@ where
             deadline: self.deadline,
         }
     }
+
+    /// See https://developer.github.com/v4/guides/resource-limitations/.
+    ///
+    /// If the query is invalid, returns `Ok(0)` so that execution proceeds and
+    /// gives a proper error.
+    pub(crate) fn root_query_complexity(
+        &self,
+        root_type: &s::TypeDefinition,
+        root_selection_set: &q::SelectionSet,
+    ) -> Result<u64, QueryExecutionError> {
+        const MAX_DEPTH: u32 = 100;
+
+        match self.query_complexity(root_type, root_selection_set, MAX_DEPTH, 0) {
+            Ok(complexity) => Ok(complexity),
+            Err(ComplexityError::Invalid) => Ok(0),
+            Err(ComplexityError::TooDeep) => Err(QueryExecutionError::TooDeep),
+            Err(ComplexityError::Overflow) => {
+                Err(QueryExecutionError::TooComplex(u64::max_value(), 0))
+            }
+        }
+    }
+
+    fn query_complexity(
+        &self,
+        ty: &s::TypeDefinition,
+        selection_set: &q::SelectionSet,
+        max_depth: u32,
+        depth: u32,
+    ) -> Result<u64, ComplexityError> {
+        use ComplexityError::*;
+
+        if depth > max_depth {
+            return Err(TooDeep);
+        }
+
+        // Helpers to look for types and fields on both the introspection and
+        // regular schemas.
+        fn get_named_type(
+            schema: &s::Document,
+            name: &Name,
+        ) -> Result<s::TypeDefinition, ComplexityError> {
+            if name.starts_with("__") {
+                let doc = introspection_schema(SubgraphDeploymentId::new("").unwrap()).document;
```

Comment:
I may be wrong but creating the introspection schema for every matching named type feels expensive.

Reply:
I wonder if this could benefit from @lutter's splitting, where the schema is decided at the top level.

Reply:
Good catch, I've put the introspection document in a `lazy_static`. This is being called before the logic for deciding the schema. Though it would be possible to call it after, then we'd sum the complexity for the introspection and regular sides.